routes/gallery.js

```diff
@@ -0,0 +1,106 @@
+const express = require('express');
+const router = express.Router();
+const knex = require('../database/knex');
+
+function isAuthenticated(req, res, next) {
+  if (req.isAuthenticated()) { next(); }
+  else {
+    res.redirect('/login');
+  }
+}
+
+router.route('/')
+  .post(isAuthenticated, function (req, res) {
+    knex('images')
+      .insert(req.body)
+      .then(function () {
+        knex.select('id')
+          .from('images')
+          .where('link', req.body.link)
+          .then(function (image) {
+            res.redirect(image[0].id);
+          });
+      })
+      .catch(function () {
+        res.redirect('gallery/new');
+      })
+  });
+
+router.route('/new')
+  .get(isAuthenticated, function (req, res) {
+    res.render('gallery/new')
+  });
+
+router.route('/:id')
+  .get(function (req, res) {
+    knex.select('author', 'link', 'description', 'id')
+      .from('images')
+      .where('id', req.params.id)
+      .then(function (image) {
+        if (image[0]) {
+          knex.select('id', 'author', 'link')
```

Comment:
This promise chain should be returned so that it could can merged into the promise above this one. Even better, if there is not image, you should throw an error, otherwise the promise chain can move on to the next `.then` which should be this promise chain.

Reply:
The same goes for all other routes below where you have two promise chains within each other but not having the nested promise returning to the parent promise. This is effectively the same as using callback and promotes the pyramid structure rather than using proper promises and having a more top down / inline structure.